File: src/p0015/cpp/solution.cpp
/**
 * Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]] such that i != j, i != k, and j != k, and nums[i] + nums[j] + nums[k] == 0.
 * 
 * Notice that the solution set must not contain duplicate triplets.
 * 
 *  
 * 
 * Example 1:
 * 
 * Input: nums = [-1,0,1,2,-1,-4]
 * Output: [[-1,-1,2],[-1,0,1]]
 * Example 2:
 * 
 * Input: nums = []
 * Output: []
 * Example 3:
 * 
 * Input: nums = [0]
 * Output: []
 *  
 * 
 * Constraints:
 * 
 * 0 <= nums.length <= 3000
 * -105 <= nums[i] <= 105
 */


class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
		sort(nums.begin(), nums.end());
		vector<vector<int>> ret;
		for(int i = 0; i < nums.size(); i++) {
			if(i > 0 && nums[i] == nums[i-1]) { continue; }
			for(int j = i+1, k=nums.size()-1; j < nums.size(); j++) {
				if(j > i+1 && nums[j] == nums[j-1]) { continue; }
				while(k > j && nums[j]+nums[k] > -1*nums[i]) { k--; }
				if(k == j) { break; }
				if(nums[i] + nums[j] + nums[k] == 0) {
					ret.push_back({nums[i], nums[j], nums[k]});
				}
			}
		}

		return ret;
    }
};

File: src/p0015/python/solution.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        ans = []
        nums.sort()
        n = len(nums)
        for k in range(n):
            if k > 0 and nums[k] == nums[k-1]: continue
            j = n - 1
            for i in range(k+1, n):
                if i > k+1 and nums[i-1] == nums[i]: continue
                while i < j and nums[k] + nums[i] + nums[j] > 0: j -= 1
                if i == j: break
                if nums[k] + nums[i] + nums[j] == 0:
                    ans.append([nums[k], nums[i], nums[j]])
            
        return ans